docs/source/development/eep-01-pytrees.rst

```diff
@@ -0,0 +1,848 @@
+
+What is a container can be re-defined for each application. By default, lists, tuples
+and dicts are considered containers and everything else is a leaf. Then the following
+are examples of pytrees:
+
+
+.. code-block:: python
+
+    [1, "a", np.arange(3)]  # 3 leaves
+
+    [1, {"k1": 2, "k2": (3, 4)}, 5]  # 5 leaves
+
+    np.arange(5)  # 1 leaf
+
+
+What makes pytrees so powerful are the operations defined for them. The most important
+ones are:
+
+- ``tree_flatten``: Convert any pytree into a flat list of leaves + metadata
+- ``tree_unflatten``: The inverse of ``tree_flatten``
+- ``tree_map``: Apply a function to all leaves in a pytree
+- ``leaf_names``: Generate a list of names for all leaves in a pytree
+
+The above examples of pytrees would look as follows when flattened (with a default
+definition of containers):
+
+.. code-block:: python
+
+    [1, "a", np.arange(3)]
+
+    [1, 2, 3, 4, 5]
+
+    [np.arange(5)]
+
+By adding numpy arrays to the registry of container like objects, each of the three
+examples above would have five leafs. The flattened versions would look as follows:
+
+.. code-block:: python
+
+    [1, "a", 0, 1, 2]
+
+    [1, 2, 3, 4, 5]
+
+    [0, 1, 2, 3, 4]
+
+Needless to say, it is possible to register any iterable as container. For example, we
+would add pandas.Series and pandas.DataFrame (with varying definitions, depending on the
+application).
+
+
+Difference between pytrees in JAX and estimagic
+===============================================
+
+Most JAX functions `only work with Pytrees of arrays
+<https://jax.readthedocs.io/en/latest/pytrees.html#pytrees-and-jax-functions>`_, i.e.
+pytrees where container types are dicts, lists and tuples and all leaves are arrays.
+
+There are two ways to look at such pytrees:
+
+1. As pytree of arrays -> ``tree_flatten`` produces a list of arrays
+2. As pytree of numbers -> ``tree_flatten`` produces a list of numbers
+
+The only difference between the two views is that for the second one, arrays have been
+registered as container types that can be flattened.
+
+Estimagic functions work with slightly more general pytrees. On top of arrays, they
+can also contain scalars, pandas.Series and pandas.DataFrames.
+
+Again, there are two possible ways to look at such pytrees:
+
+1. As pytree of arrays, numbers, Series and DataFrames -> ``tree_flatten`` produces a
+   list of arrays numbers, Series and DataFrames.
+2. As pytree of numbers -> ``tree_flatten`` produces a list of numbers
+
+Again, the difference between the two is which objects are registered as container types
+and the rules to flatten and unflatten them are defined.
+
+While numpy arrays, scalars and pandas.Series have only one natural way of defining
+the flattening rules, this becomes more complex for DataFrames due to the way ``params``
+DataFrames were used in estimagic before.
+
+We define the following rules: If a DataFrame contains a column called ``"value"``, we
```

Comment:
My initial reaction was: What happens with bounds, then?

But really that thought should not arise. Thus make clearer somewhere what these guys are used for: For specifying start values / evaluations / etc., so the issue of flattening both dimensions of a DataFrame never arises, essentially only ever interested in a single Series of that DataFrame.

Reply:
Made it clearer.